File: LeetCode/98. Validate Binary Search Tree.py
# LeetCode 98. Validate Binary Search Tree
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.checkBST(root)
    
    def checkBST(self, root : TreeNode, low = -math.inf, high = math.inf) : 
        if not root : return True
        if root.val <= low or root.val >= high : return False
        return self.checkBST(root.left, low, root.val) and self.checkBST(root.right, root.val, high)

File: LeetCode/98. Validate Binary Search Tree.cpp
// LeetCode 98. Validate Binary Search Tree
// DFS.
class Solution{
public:
    bool helper(TreeNode *root, long long low, long high){
        if(root == NULL) return true;
        if(root -> val <= low || high <= root -> val) return false;
        return helper(root -> left, low, root -> val) && helper(root -> right, root -> val, high);
    }

    bool isValidBST(TreeNode* root){
        // LONG_LONG_MIN & LONG_LONG_MAX exist too.
        return helper(root, LONG_MIN, LONG_MAX);
    }
};
